viewing_party/party.py

```diff
@@ -1,23 +1,257 @@
+    num_of_movies = len(user_data["watched"])
+    return total_rating / num_of_movies
+
+
+def get_most_watched_genre(user_data):
+    genres_most_watched = {}
+
+    # Iterate through "watched" list
+    for i in range(len(user_data['watched'])):
+        genre = user_data['watched'][i]['genre']
+        if genre in genres_most_watched:
+            genres_most_watched[genre] += 1
+        else:
+            genres_most_watched[genre] = 1
+
+    highest_occurence = 0
+    most_watched = None
+
+    # Iterate through out genres_most_watched dict
+    for genre, num in genres_most_watched.items():
+        if num > highest_occurence:
+            highest_occurence = num
+            most_watched = genre
+            
+    return most_watched
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+
+    copy_from_user_data = copy.deepcopy(user_data)
+
+    friends_movie_titles = []
+    users_movie_titles = []
+    my_list = copy_from_user_data["watched"]
+    friends_list = copy_from_user_data["friends"]
+    
+    for i in range(len(friends_list)):
+        values_of_watched = friends_list[i]["watched"]
         
+        for j in range(len(values_of_watched)):
+            titles = values_of_watched[j]["title"]
+            friends_movie_titles.append(titles)
+
+    for i in range(len(my_list)):
+        users_movie_titles.append(my_list[i]["title"])
+
+    user_unique_movies = set(users_movie_titles) - set(friends_movie_titles)
+    users_unique_movies_list = list(user_unique_movies)
+    
+    result_list = []
+
+    for movie_title in users_unique_movies_list:
+        for movie in my_list:
+            if movie["title"] == movie_title:
+                result_list.append(movie)
+
+    return result_list
+
+
+def get_friends_unique_watched(user_data):
+
+    copy_from_user_data = copy.deepcopy(user_data)
+    
+    friends_movie_titles = []
+    users_movie_titles = []
+    friends_movies = []
+    my_list = copy_from_user_data["watched"]
+    friends_list = copy_from_user_data["friends"]
+    
+    for i in range(len(friends_list)):
+        values_of_watched = friends_list[i]["watched"]
+        
+        for j in range(len(values_of_watched)):
+            titles = values_of_watched[j]["title"]
+            friends_movie_titles.append(titles)
+            friends_movies.append(values_of_watched[j])
+
+    for i in range(len(my_list)):
+        users_movie_titles.append(my_list[i]["title"])
+
+    friends_unique_movies = set(friends_movie_titles) - set(users_movie_titles)
+    friends_unique_movies_list = list(friends_unique_movies)
+    
+    result_list = []
+
+    for movie_title in friends_unique_movies_list:
+        for movie in friends_movies:
+            if movie["title"] == movie_title:
+                result_list.append(movie) 
+                break
+
+    return result_list

```

Comment:
👍  Nice work! Sets are really a great tool for storing immutable data. However, could we have found the friend's unique movies without the set and the extra iteration?  There are always tradeoffs when we use a specific data structure and when creating a solution we should try to consider the added time/space complexity of using that data structure.

 When using the set solution for this function, it adds the extra step of creating the set and then having to iterate over it again to create another data structure (`result_list`). What would this solution look like if we hadn't used the set? What would this solution look like if we used a helper function to find the friend's movies?